### PR TITLE
[BEAM-1258] DONT_RETRY_NOT_FOUND in BigQueryServicesImpl.isTableEmpty().

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi

**Comment by dhalperi** [comment]
Please add a new unit test to catch the regression here.

**Comment by dhalperi** [comment]
Also, I think there is also a problem in `getOrCreateTable`:

```
Exception: java.io.IOException: Unable to get table: streamdemo, aborting after 9 retries.
org.apache.beam.sdk.io.gcp.bigquery.BigQueryServicesImpl.executeWithRetries(BigQueryServicesImpl.java:869)
org.apache.beam.sdk.io.gcp.bigquery.BigQueryServicesImpl$DatasetServiceImpl.getTable(BigQueryServicesImpl.java:401)
org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$StreamingWriteFn.getOrCreateTable(BigQueryIO.java:2658)
org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$StreamingWriteFn.finishBundle(BigQueryIO.java:2632)
Caused By: com.google.api.client.googleapis.json.GoogleJsonResponseException: 404 Not Found
{
  "code" : 404,
  "errors" : [ {
    "domain" : "global",
    "message" : "Not found: Table ",
    "reason" : "notFound"
  } ],
  "message" : "Not found: Table "
}
com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113)
com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40)
com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:321)
com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1065)
```

**Comment by peihe** [comment]
PTAL

Added tests.

getTable() fix is sent as a separate PR

**Comment by dhalperi** [comment]
Looks like it needs a rebase -- sorry. May have merged this and #1760 in the wrong order.

**Comment by peihe** [comment]
PTAL

rebased

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java [hunk 1]
```diff
@@ -385,6 +387,78 @@ public void testGetTableThrows() throws Exception {
...
+    TableReference tableRef = new TableReference()
+        .setProjectId("projectId")
+        .setDatasetId("datasetId")
+        .setTableId("tableId");
+
+    thrown.expect(IOException.class);
+    thrown.expectMessage(String.format("Unable to list table data: %s", tableRef.getTableId()));
+
+    BigQueryServicesImpl.DatasetServiceImpl datasetService =
+        new BigQueryServicesImpl.DatasetServiceImpl(bigquery, PipelineOptionsFactory.create());
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java]: Configure `thrown` as late as possible. Aka, right before L458, after the setup of datasetService.

**Comment by dhalperi** [comment]
retest this please

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java (hunk 1 above)
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImplTest.java]: done

**Comment by dhalperi** [comment]
retest this please

**Pull request closed without merging**